paka/kube_resources/model_group/models/abstract.py

```diff
@@ -0,0 +1,338 @@
+            download_max_concurrency (int, optional): The maximum number of concurrent downloads. Defaults to 10.
+            s3_chunk_size (int, optional): The size of each chunk to upload to S3 in bytes. Defaults to 8 * 1024 * 1024.
+            s3_max_concurrency (int, optional): The maximum number of concurrent uploads to S3. Defaults to 20.
+        """
+        # model info
+        self.name = name
+        self.inference_devices = inference_devices
+        self.quantization = quantization
+        self.runtime = runtime
+        self.prompt_template = prompt_template
+
+        # s3 bucket
+        self.s3_bucket = read_current_cluster_data("bucket")
+        self.s3_chunk_size = s3_chunk_size
+        self.download_max_concurrency = download_max_concurrency
+        self.s3_max_concurrency = s3_max_concurrency
+        self.s3 = boto3.client("s3", config=Config(signature_version="s3v4"))
+
+    def get_s3_file_path(self, file_path: str) -> str:
+        """
+        Returns the S3 file path for a given file name.
+
+        Args:
+            file_path (str): The path of the file.
+
+        Returns:
+            str: The S3 file path.
+        """
+        return f"{MODEL_PATH_PREFIX}/{file_path}"
+
+    def download(self, url: str, sha256: str | None = None) -> None:
+        """
+        Downloads a single file from a URL.
+
+        Args:
+            url (str): The URL of the file to download.
+            sha256 (str, optional): The expected SHA256 hash of the downloaded file.
+
+        Raises:
+            Exception: If the SHA256 hash of the downloaded file does not match the expected value.
+        """
+        full_model_file_path = self.get_s3_file_path(
+            f"{self.name}/{url.split('/')[-1]}"
+        )
+        if self.s3_file_exists(full_model_file_path):
+            logger.info(f"Model file {full_model_file_path} already exists.")
+            return
+
+        logger.info(f"Downloading model from {url}")
+        completed_upload_id = None
+        try:
+            with requests.get(url, stream=True) as response:
+                response.raise_for_status()
+                upload_id, sha256_value = self.upload_to_s3(
+                    response, full_model_file_path
+                )
+                if sha256 is not None and sha256 != sha256_value:
+                    raise Exception(
+                        f"SHA256 hash of the downloaded file does not match the expected value. {full_model_file_path}"
+                    )
+                completed_upload_id = upload_id
+        except Exception as e:
+            logger.error(f"An error occurred, download: {str(e)}")
+            raise e
+        finally:
+            # If an error occurred and upload was not completed
+            if completed_upload_id is None:
+                self.s3.abort_multipart_upload(
+                    Bucket=self.s3_bucket, Key=full_model_file_path, UploadId=upload_id
+                )
```

Comment:
this doesn't seem to be right. If completed_upload_id is None, most likely upload_id is None and you aborting a None upload id. It is better that you cancel the mutliplepart_upload in "upload_to_s3". otherwise, it is hard for you handle this scenario, upload_id is created (not None), but a part upload failed.